internal/langserver/handlers/tfschema/candidates.go

```diff
@@ -60,10 +60,24 @@ func valueCandidates(values []string, r lsp.Range, isOrdered bool) []lsp.Complet
 	return candidates
 }
 
-func typeCandidates(prefix *string, r lsp.Range) []lsp.CompletionItem {
+func resourceTypeCandidates(prefix *string, r lsp.Range) []lsp.CompletionItem {
+	return typeCandidates(prefix, r, false)
+}
+
+func dataSourceTypeCandidates(prefix *string, r lsp.Range) []lsp.CompletionItem {
+	return typeCandidates(prefix, r, true)
+}
+
+func typeCandidates(prefix *string, r lsp.Range, allowReadOnly bool) []lsp.CompletionItem {
 	candidates := make([]lsp.CompletionItem, 0)
 	if prefix == nil || !strings.Contains(*prefix, "@") {
-		for resourceType := range azure.GetAzureSchema().Resources {
+		for resourceType, resource := range azure.GetAzureSchema().Resources {
+			if len(resource.Definitions) > 0 {
+				def, err := resource.Definitions[0].GetDefinition()
+				if err == nil && def.IsReadOnly() && !allowReadOnly {
```

Comment:
What if `err` is not nil here?

Reply:
Normally err is always nil here, but if err is not nil, then we can't know whether this type is readonly or not, so skip checking on it.